src/two_finger_grasp_generator.cpp

```diff
@@ -0,0 +1,886 @@
+  if (!grasp_data->setGraspWidth(percent_open, min_finger_open_on_approach, new_grasp.pre_grasp_posture))
+  {
+    debugFailedOpenGripper(percent_open, min_finger_open_on_approach, object_width,
+                           grasp_data->grasp_padding_on_approach_);
+    return false;
+  }
+  new_grasp.grasp_quality = scoreFingerGrasp(grasp_pose_tcp, grasp_data, object_pose, percent_open);
+  grasp_candidates.push_back(std::make_shared<GraspCandidate>(new_grasp, grasp_data, object_pose));
+
+  return true;
+}
+
+bool TwoFingerGraspGenerator::generateCuboidAxisGrasps(const Eigen::Isometry3d& cuboid_pose, double depth, double width,
+                                                       double height, grasp_axis_t axis,
+                                                       const TwoFingerGraspDataPtr& grasp_data,
+                                                       const TwoFingerGraspCandidateConfig& grasp_candidate_config,
+                                                       std::vector<GraspCandidatePtr>& grasp_candidates)
+{
+  double finger_depth = grasp_data->grasp_max_depth_ - grasp_data->grasp_min_depth_;
+  double length_along_a, length_along_b, length_along_c;
+  double delta_a, delta_b, delta_f;
+  double alpha_x, alpha_y, alpha_z;
+  Eigen::Vector3d object_size(depth, width, height);
+
+  double object_width;
+  EigenSTL::vector_Isometry3d grasp_poses_tcp;
+
+  Eigen::Isometry3d grasp_pose_tcp = cuboid_pose;
+  Eigen::Vector3d a_dir, b_dir, c_dir;
+
+  if (axis == X_AXIS)
+  {
+    length_along_a = width;
+    length_along_b = height;
+    length_along_c = depth;
+    a_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitY();
+    b_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitZ();
+    c_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitX();
+    alpha_x = -M_PI / 2.0;
+    alpha_y = 0;
+    alpha_z = -M_PI / 2.0;
+    object_width = depth;
+  }
+  else if (axis == Y_AXIS)
+  {
+    length_along_a = depth;
+    length_along_b = height;
+    length_along_c = width;
+    a_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitX();
+    b_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitZ();
+    c_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitY();
+    alpha_x = 0;
+    alpha_y = M_PI / 2.0;
+    alpha_z = M_PI;
+    object_width = width;
+  }
+  else  // Z_AXIS
+  {
+    length_along_a = depth;
+    length_along_b = width;
+    length_along_c = height;
+    a_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitX();
+    b_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitY();
+    c_dir = grasp_pose_tcp.rotation() * Eigen::Vector3d::UnitZ();
+    alpha_x = M_PI / 2.0;
+    alpha_y = M_PI / 2.0;
+    alpha_z = 0;
+    object_width = height;
+  }
+
+  double rotation_angles[3];
+  rotation_angles[0] = alpha_x;
+  rotation_angles[1] = alpha_y;
+  rotation_angles[2] = alpha_z;
+
+  a_dir = a_dir.normalized();
+  b_dir = b_dir.normalized();
+  c_dir = c_dir.normalized();
```

Comment:
There's a `normalize()` in place operation for Eigen right?  I don't think aliasing is a problem here but it might be safer.

Reply:
you are right but I'd rather not touch this code too much